doc/man/pilot-read-todos.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "pilot-read-todos" 1 "Copyright 1996-2007 FSF" "0.12.5" "PILOT-LINK"
.SH NAME
pilot-read-todos \- Syncronize your Palm ToDo application's database with your desktop machine.
.SH "SECTION"

.PP
pilot\-link: Userland conduits

.SH "SYNOPSIS"

.PP
 \fIpilot\-read\-todos\fR [\fB\-p\fR|\fB\-\-port\fR <\fBport\fR>] [\fB\-q\fR|\fB\-\-quiet\fR] [\fB\-A\fR|\fB\-\-archived\fR] [\fB\-w\fR|\fB\-\-write\fR]

.PP
 \fIpilot\-read\-todos\fR [\fB\-f\fR|\fB\-\-file\fR  \fIfile\fR] [\fB\-A\fR|\fB\-\-archived\fR] [\fB\-w\fR|\fB\-\-write\fR]

.PP
 \fIpilot\-read\-todos\fR [\fB\-\-version\fR] [\fB\-?\fR|\fB\-\-help\fR] [\fB\-\-usage\fR]

.SH "DESCRIPTION"

.PP
 \fIpilot\-read\-todos\fR will read the ToDo database on the Palm handheld, or read the ToDoDB\&.pdb file directly from disk on your desktop machine, and print the contents to STDOUT (generally the screen)\&.

.PP
Use only one of \fB\-\-port\fR or \fB\-\-file\fR\&.

.SH "OPTIONS"

.SS "pilot-read-todos option"

                        \fB\-f\fR, \fB\-\-file\fR \fIfile\fR
                    
.PP
Read a local todo database file from disk for printing\&. Excludes \-\-port\&.

                        \fB\-A\fR, \fB\-\-archived\fR
                    
.PP
Write archived entries only, in human\-readable format

                        \fB\-w\fR, \fB\-\-write\fR
                    
.PP
Write output \- every operation must include a write command\&.

.SS "Conduit Options"

                        \fB\-p\fR, \fB\-\-port\fR \fIport\fR
                    
.PP
Use device file <\fIport\fR> to communicate with the Palm handheld device\&. If this is not specified, \fIpilot\-read\-todos\fR will look for the \fI $PILOTPORT \fR environment variable\&. If neither are found or supplied, \fIpilot\-read\-todos\fR will print the usage information\&.

.PP
Excludes \-\-file\&.

                        \fB\-q\fR, 
                        \fB\-\-quiet\fR
                    
.PP
Suppress 'Hit HotSync button' message

                        \fB\-v\fR, \fB\-\-version\fR
                    
.PP
Display version of \fIpilot\-read\-todos\fR and exit without connecting\&.

.SS "Help Options"

                        \fB\-h\fR, \fB\-\-help\fR
                    
.PP
Display the help synopsis for \fIpilot\-read\-todos\fR and exit without connecting\&.

                        \fB\-\-usage\fR 
                    
.PP
Display a brief usage message and exit without connecting\&.

.SH "USAGE"

.PP
Connect to a target Palm handheld and read the ToDo database, or read the ToDoDB\&.pdb file from disk and print the entries found to STDOUT\&.

.SH "EXAMPLES"

.PP
Read the on\-Palm ToDo database, and dump to STDOUT\&.

.PP
 \fIpilot\-read\-todos\fR  \fB\-w\fR  \fB\-p\fR  \fI/dev/pilot\fR 

.PP
Read the ToDoDB\&.pdb file on disk, and dump to STDOUT

.PP
 \fIpilot\-read\-todos\fR  \fB\-w\fR  \fB\-f\fR  \fIToDoDB\&.pdb\fR 

.SH "AUTHOR"

.PP
 \fIpilot\-read\-todos\fR was originally written by Kenneth Albanowski

.SH "BUGS"

.SS "Known Bugs"

.PP
 \fIpilot\-read\-todos\fR does not support Notes attached to ToDo items at this time\&.

.SS "Reporting Bugs"

.PP
We have an online bug tracker\&. Using this is the only way to ensure that your bugs are recorded and that we can track them until they are resolved or closed\&. Reporting bugs via email, while easy, is not very useful in terms of accountability\&. Please point your browser to http://bugs\&.pilot\-link\&.org: \fIhttp://bugs.pilot-link.org\fR and report your bugs and issues there\&.

.SH "COPYRIGHT"

.PP
This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version\&.

.PP
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\&. See the GNU General Public License for more details\&.

.PP
You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc\&., 51 Franklin St, Fifth Floor, Boston, MA 02110\-1301, USA\&.

.SH "SEE ALSO"

.PP
 \fIinstall\-todo\fR (1), \fIinstall\-todos\fR (1), \fIpilot\-link\fR(7)



src/pilot-read-todos.c
/*
 * $Id: pilot-read-todos.c,v 1.60 2009/06/04 13:32:32 desrod Exp $ 
 *
 * pilot-read-todos.c:  Translate Palm ToDo database into generic format
 *
 * Copyright (c) 1996, Kenneth Albanowski
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 *
 */

/* 12-27-2003:
   FIXME: Add "Private" and "Delete" flags */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "pi-socket.h"
#include "pi-todo.h"
#include "pi-file.h"
#include "pi-header.h"
#include "pi-userland.h"

void print_unarchived(struct ToDoAppInfo *tai, struct ToDo *todo, int category)
{
	printf("Category: %s\n", tai->category.name[category]);
	printf("Priority: %d\n", todo->priority);
	printf("Completed: %s\n", todo->complete ? "Yes" : "No");
	if (todo->indefinite) {
		printf("Due: No Date\n");
	} else {
		printf("Due: %s\n", asctime(&todo->due));
	}
	if (todo->description)
		printf("Description: %s\n", todo->description);
	if (todo->note)
		printf("Note: %s\n", todo->note);
	printf("\n");
}

void print_archived(struct ToDoAppInfo *tai, struct ToDo *todo, int category)
{
	printf("\"Category\", ");
	printf("\"%s\", ", tai->category.name[category]);
	printf("\"Priority\", ");
	printf("\"%d\", ", todo->priority);
	printf("\"Completed\", ");
	printf("\"%s\", ", todo->complete ? "Yes" : "No");

	if (todo->indefinite) {
		printf("\"Due\", \"No Date\", ");
	} else {
		printf("\"Due\", ");
		printf("\"%s\", ", asctime(&todo->due));
	}

	if (todo->description) {
		printf("\"Description\", ");
		printf("\"%s\", ", todo->description);
	}

	if (todo->note) {
		printf("\"Note\", ");
		printf("\"%s\", ", todo->note);
	}

	printf("\n\n");
}

int main(int argc, const char *argv[])
{
	int 	c,		/* switch */
		db,
		i,
		sd 		= -1;

	enum { mode_none, mode_write=257 } run_mode = mode_none;
	int archived = 0;

	char
		*filename 	= NULL,
		*ptr;

	struct 	PilotUser User;
	struct 	pi_file *pif 	= NULL;
	struct 	ToDoAppInfo tai;

	pi_buffer_t    *recbuf,
        *appblock;

	poptContext po;

	struct poptOption options[] = {
		USERLAND_RESERVED_OPTIONS
		{"write",	'w', POPT_ARG_NONE,NULL,mode_write, "Write output" },
	        {"file", 	'f', POPT_ARG_STRING, &filename, 0, "Save ToDO entries in <filename> instead of STDOUT"},
		{"archived",	'A', POPT_ARG_NONE, &archived, 0, "Write archived entries only, in human-readable format"},
		POPT_TABLEEND
	};

	po = poptGetContext("read-todos", argc, argv, options, 0);
	poptSetOtherOptionHelp(po,"\n\n"
	"   Synchronize your ToDo database with your desktop machine.\n"
	"   If you use --port, the contents of your Palm's ToDo database will be written to\n"
	"   standard output in a generic text format/ Otherwise, use --file to read a todo\n"
	"   database file from disk for printing.\n\n"
	"   Example arguments:\n"
	"      -w -A -p /dev/pilot \n"
	"      -w -f ToDoDB.pdb\n");


	if (argc<2) {
		poptPrintUsage(po,stderr,0);
		return 1;
	}

	while ((c = poptGetNextOpt(po)) >= 0) {
		switch(c) {
		case mode_write :
			if (run_mode == mode_none) {
				run_mode = c;
			} else {
				if (c!=run_mode) {
					fprintf(stderr,"   ERROR: Specify exactly one of -w.\n");
					return 1;
				}
			}
			break;
		default:
			fprintf(stderr,"   ERROR: Unhandled option %d.\n",c);
			return 1;
		}
	}

	if (c < -1) {
		plu_badoption(po,c);
	}

	if (mode_none == run_mode) {
		fprintf(stderr,"   ERROR: Specify exactly one of -w.\n");
		return 1;
	}
	if (!plu_port && !filename) {
		fprintf(stderr,"   ERROR: Specify either --port or --file.\n");
		return 1;
	}

	/* Read ToDoDB.pdb from the Palm directly */
    appblock = pi_buffer_new(0xffff);
	if (!filename) {
		sd = plu_connect();
		if (sd < 0)
			goto error;

		if (dlp_ReadUserInfo(sd, &User) < 0)
			goto error_close;

		/* Open the ToDo database, store access handle in db */
		if (dlp_OpenDB(sd, 0, 0x80 | 0x40, "ToDoDB", &db) < 0) {
			puts("   Unable to open ToDoDB");
			dlp_AddSyncLogEntry(sd,
					    "   Unable to open ToDoDB.\n");
			exit(EXIT_FAILURE);
		}

		dlp_ReadAppBlock(sd, db, 0, appblock->allocated, appblock);

	/* Read ToDoDB.pdb from disk */
	} else {
		size_t 	len;

		pif = pi_file_open(filename);

		if (!pif) {
			fprintf(stderr, "   ERROR: %s\n", strerror(errno));
			fprintf(stderr, "   Does %s exist?\n\n", filename);
			exit(EXIT_FAILURE);
		}

		pi_file_get_app_info(pif, (void *) &ptr, &len);
        pi_buffer_append(appblock, ptr, len);
	}

	unpack_ToDoAppInfo(&tai, appblock->data, appblock->used);
    pi_buffer_free(appblock);

	recbuf = pi_buffer_new (0xffff);

	for (i = 0;; i++) {
		int 	attr,
			category,
			len;

		struct 	ToDo todo;

		if (!filename) {
			len =
			    dlp_ReadRecordByIndex(sd, db, i, recbuf, 0,
						  &attr, &category);

			if (len < 0)
				break;
		}
		else {
			if (pi_file_read_record
			    (pif, i, (void *) &ptr, &len, &attr, &category,
			     0))
				break;

			pi_buffer_clear(recbuf);
			pi_buffer_append(recbuf, ptr, len);
		}

		/* Skip deleted records */
		if (attr & dlpRecAttrArchived) {
			if (archived) {
				unpack_ToDo(&todo, recbuf, todo_v1);
				print_archived(&tai,&todo,category);
				free_ToDo(&todo);
			}
			continue;
		}
		if (attr & dlpRecAttrDeleted)
			continue;

		if (!archived) {
			unpack_ToDo(&todo, recbuf, todo_v1);
			print_unarchived(&tai,&todo,category);
			free_ToDo(&todo);
		}
	}

    pi_buffer_free (recbuf);

	if (!filename) {
		/* Close the database */
		dlp_CloseDB(sd, db);
		dlp_AddSyncLogEntry(sd, "Successfully read ToDos from Palm.\n"
					"Thank you for using pilot-link.");
		dlp_EndOfSync(sd, 0);
		pi_close(sd);

	} else {
		pi_file_close(pif);
	}
	return 0;

error_close:
	pi_close(sd);

error:
	return -1;
}

/* vi: set ts=8 sw=4 sts=4 noexpandtab: cin */
/* ex: set tabstop=4 expandtab: */
/* Local Variables: */
/* indent-tabs-mode: t */
/* c-basic-offset: 8 */
/* End: */
